### ISSUE TITLE
Creation of Named Constraints in Disjunctions

### ISSUE THREAD

**Issue by iamrecursion** [LeventErkok/sbv#372]
I've had to change some of the implementation of my solver using sbv, and I'm wondering if it's possible to somehow name individual elements of a disjunction so as to provide more informative error messages via the unsat core. 

Consider a set of constraints `c1 = [v1, v2, v3]`, only one of which can be true at a given time. You can do the following, but it only lets you name the resultant disjunction:

```hs
namedConstraint "myDisjunction" (bOr c1)
```
This is fine if each `vn` is a simple constraint by itself, but if they are composed of multiple constraints  (disjunctive normal form) it would be beneficial to name those sub-constraints (the conjunctions) to know which one(s) contribute to the unsatisfiability of the formula as a whole. Is this possible currently?

**Comment by LeventErkok** [comment]
I don't think this is possible in SMTLib. Page 67 of http://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2017-07-18.pdf says (at the very top):

> (get-unsat-core) asks the solver to identify an unsatisfiable core, a subset of all the formulas
> in the current context that is unsatisfiable by itself. The command can be issued only
> if the most recent check command had an empty set of assumptions. The solver selects
> from the unsatisfiable core only those formulas that have been asserted with a command
> of the form (assert (! t :named f )), and returns a sequence ( f1 · · · fn) of those
> labels. Unlabeled formulas in the unsatisfiable core are simply not reported.(51)
> The semantics of this command’s output is that the reported assertions together with all
> the unlabeled ones in the set of all assertions are jointly unsatisfiable. In practice then,
> not labeling assertions is useful for unsat core detection purposes only when the user is
> sure that the set of all unlabeled assertions is satisfiable.
> The command can be issued only if the :produce-unsat-cores option, which is set to
> false by default, is set to true (see Section 4.1.7).

And this is essentially the mechanism SBV uses to translate `namedConstraint`.

Regardless of SBV, if you can figure out a way to do what you want using SMTLib, then we can see if there's a way to generate the corresponding code from SBV. From the above paragraph though, I don't see how to name something without also asserting it.

**Comment by iamrecursion** [comment]
I didn't even think to check the SMTLib spec. Having just given the relevant sections a skim, it doesn't look like there's a way to do what I want as, like you've mentioned, you seemingly can't name anything without asserting it. Unfortunately there seems to be no way to perform conditional assertions, and hence no way to solve this problem!

**Comment by LeventErkok** [comment]
The other alternative would be to loop over your `c` set and assert one at a time after you first get the unsat; but of course that'd require multiple calls.

**Comment by LeventErkok** [comment]
Can we close this ticket? I don't think there's much SBV can do regarding this; due to how unsat-cores work.
